RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitTimingCCAlgo.cc

```diff
@@ -0,0 +1,167 @@
+      gainratio = aGain->gain12Over6();
+    }
+
+    amplitude = ((double)(sample.adc()) - pedestal) * gainratio;
+
+    if (gainId == 0) {
+      //saturation
+      amplitude = (4095. - pedestal) * gainratio;
+    }
+
+    pedSubSamples.at(iSample) = amplitude;
+
+    if (amplitude > maxamplitude) {
+      maxamplitude = amplitude;
+    }
+    pulsenorm += fullpulse(iSample);
+  }
+
+  std::vector<double>::const_iterator amplit;
+  for (amplit = amplitudes.begin(); amplit < amplitudes.end(); ++amplit) {
+    int ipulse = std::distance(amplitudes.begin(), amplit);
+    // The following 3 lines are copied from EcalRecAlgos/interface/EcalUncalibRecHitTimeWeightsAlgo.h
+    int bx = ipulse - 5;
+    int firstsamplet = std::max(0, bx + 3);
+    int offset = 7 - 3 - bx;
+
+    std::vector<double> pulse(nsample);
+    for (unsigned int isample = firstsamplet; isample < nsample; ++isample) {
+      pulse.at(isample) = fullpulse(isample + offset);
+      pedSubSamples.at(isample) =
+          std::max(0., pedSubSamples.at(isample) - amplitudes[ipulse] * pulse.at(isample) / pulsenorm);
+    }
+  }
+
+  float tStart = startTime_ + GLOBAL_TIME_SHIFT;
+  float tStop = stopTime_ + GLOBAL_TIME_SHIFT;
+  float tM = (tStart + tStop) / 2;
+
+  float distStart, distStop;
+  int counter = 0;
+
+  do {
+    ++counter;
+    distStart = computeCC(pedSubSamples, fullpulse, tStart);
+    distStop = computeCC(pedSubSamples, fullpulse, tStop);
+
+    if (distStart > distStop) {
+      tStop = tM;
+    } else {
+      tStart = tM;
+    }
+    tM = (tStart + tStop) / 2;
+
+  } while (tStop - tStart > targetTimePrecision_ && counter < MAX_NUM_OF_ITERATIONS);
+
+  tM -= GLOBAL_TIME_SHIFT;
+
+  if (counter < MIN_NUM_OF_ITERATIONS || counter > MAX_NUM_OF_ITERATIONS - 1) {
+    if (counter > MAX_NUM_OF_ITERATIONS / 2)
+      //Produce a log if minimization took too long
+      edm::LogInfo("EcalUncalibRecHitTimingCCAlgo::computeTimeCC")
+          << "Minimization Counter too high: " << counter << std::endl;
+    tM = TIME_WHEN_NOT_CONVERGING * ecalPh1::Samp_Period;
+    //Negative error means that there was a problem with the CC
+    errOnTime = -targetTimePrecision_ / ecalPh1::Samp_Period;
+  }
+
+  return -tM / ecalPh1::Samp_Period;
+}
+
+FullSampleVector EcalUncalibRecHitTimingCCAlgo::interpolatePulse(const FullSampleVector& fullpulse, const float t) {
+  // t is in ns
+  int shift = t / ecalPh1::Samp_Period;
+  if (t < 0)
+    shift -= 1;
+  float timeShift = t - ecalPh1::Samp_Period * shift;
+  float tt = timeShift / ecalPh1::Samp_Period;
+
+  FullSampleVector interpPulse;
+  // 2nd poly with avg
+  unsigned int numberOfSamples = fullpulse.size();
+  for (unsigned int i = 1; i < numberOfSamples - 2; ++i) {
+    float a = 0.25 * tt * (tt - 1) * fullpulse[i - 1] + (0.25 * (tt - 2) - 0.5 * (tt + 1)) * (tt - 1) * fullpulse[i] +
+              (0.25 * (tt + 1) - 0.5 * (tt - 2)) * tt * fullpulse[i + 1] + 0.25 * (tt - 1) * tt * fullpulse[i + 2];
```

Comment:
```suggestion
  auto facM1orP2 = 0.25 * tt * (tt - 1);
  auto fac = (0.25 * (tt - 2) - 0.5 * (tt + 1)) * (tt - 1);
  auto facP1 = (0.25 * (tt + 1) - 0.5 * (tt - 2)) * tt;
  for (unsigned int i = 1; i < numberOfSamples - 2; ++i) {
    float a = facM1orP2 * fullpulse[i - 1] + fac * fullpulse[i] +
              facP1 * fullpulse[i + 1] + facM1orP2 * fullpulse[i + 2];
```
to avoid repeated recomputations inside the loop